### PR TITLE
Added throttling guidance

### PR DESCRIPTION
**Pull request by johngossman**

### CODE REVIEW

**Review by cleemullins — APPROVED** [review]

**Comment by cleemullins** [comment]
@johngossman I quite like this.

I *suspect* the 0.5s P99 case may be unrealistic as a bar for using long running operations. Just setting the bar though makes architectural discussions as "This is where the bar should be" much easier.

File: Guidelines.md [hunk 1]
```diff
@@ -1623,8 +1670,82 @@ A recommended minimum retention time is 24 hours.
...
+#### 14.4.2. Rate Limits and Quotas
+When a caller has made too many calls
+
+1. Services MUST return a 429 code
+2. Services MUST return a standard error response describing the specifics so that a programmer can make appropriate changes
+3. Services MUST return a Retry-After header that indicates how long clients should wait before retrying
+4. Services MAY return RateLimit headers that document the limit or quota that has been exceeded
+5. Services MAY return RateLimit-Limit: the number of calls the client is allowed to make in a time window
+6. Services MAY return RateLimit-Remaining: the number of calls remaining in the time window
+7. Services MAY return RateLimit-Reset: the number of calls the client is allowed to make in a time window
```
Comment (daspek) [inline Guidelines.md:1718]: typo? dupe of `RateLimit-Limit`

File: Guidelines.md [hunk 2]
```diff
@@ -1623,8 +1670,82 @@ A recommended minimum retention time is 24 hours.
...
+### 14.1. Principles
+Services should be as responsive as possible, so as not to block callers. As a rule of thumb any API call that is expected to take longer than 0.5 seconds in the 99th percentile, should consider using the Long-running Operations pattern for those calls. Obviously, services cannot guarantee these response times in the face of potentially unlimited load from callers. Services should therefore design and document call request limits for clients, and respond with appropriate, actionable errors and error messages if these limits are exceeded. Services should respond quickly with an error when they are generally overloaded, rather than simply respond slowly. Finally, many services will have quotas on calls, perhaps a number of operations per hour or day, usually related to a service plan or price. When these quotas are exceeded services must also provide immediate, actionable errors. Quotas and Limits should be scoped to a customer unit: a subscription, a tenant, an application, a plan, or without any other identification a range of ip addresses…as appropriate to the service goals so that the load is properly shared and one unit is not interfering with another.
```
Comment (daspek) [inline Guidelines.md]: this document generally follows the one-sentence-per-line convention, which makes diffs easier to parse in the future. can you please put sentences on their own lines?

File: Guidelines.md [hunk 3]
```diff
@@ -1623,8 +1670,82 @@ A recommended minimum retention time is 24 hours.
...
+
+### 14.2. Return Codes (429 vs 503)
+HTTP specifies two return codes for these scenarios: ‘429 Too Many Requests’ and ‘503 Service Unavailable’. Services should use 429 for cases where clients are making too many calls and can fix the situation by changing their call pattern. Services should respond with 503 in cases where general load or other problems outside the control of the individual callers is responsible for the service becoming slow. In all cases, services should also provide information suggesting how long the callers should wait before trying in again. Clients should respect these headers and also implement other transient fault handling techniques. However, there may be clients that simply retry immediately upon failure, potentially increasing the load on the service. To handle this, services should design so that returning 429 or 503 is as inexpensive as possible, either by putting in special fastpath code, or ideally by depending on a common frontdoor or load balancer that provides this functionality.
```
Comment (daspek) [inline Guidelines.md]: replace smart quotes with '

**Review by daspek — APPROVED** [review]

File: Guidelines.md (hunk 1 above)
Comment (johngossman) [reply Guidelines.md:1718]: Nope. There are 3 RateLimit headers (and other optional ones).

File: Guidelines.md (hunk 2 above)
Comment (johngossman) [reply Guidelines.md]: Fixed

File: Guidelines.md (hunk 3 above)
Comment (johngossman) [reply Guidelines.md]: Fixed

**Comment by mackiecarr89** [comment]
Unsubscribe

---- On Mon, 01 Oct 2018 18:23:31 -0700 notifications@github.com  wrote ----

@johngossman commented on this pull request.

In Guidelines.md:

> @@ -1623,8 +1670,82 @@ A recommended minimum retention time is 24 hours.
 
 Operations SHOULD transition to "tombstone" for an additional period of time prior to being purged from the system.
 
-## 14 Push notifications via webhooks
-### 14.1 Scope
+## 14. Throttling, Quotas, and Limits
+### 14.1. Principles
+Services should be as responsive as possible, so as not to block callers. As a rule of thumb any API call that is expected to take longer than 0.5 seconds in the 99th percentile, should consider using the Long-running Operations pattern for those calls. Obviously, services cannot guarantee these response times in the face of potentially unlimited load from callers. Services should therefore design and document call request limits for clients, and respond with appropriate, actionable errors and error messages if these limits are exceeded. Services should respond quickly with an error when they are generally overloaded, rather than simply respond slowly. Finally, many services will have quotas on calls, perhaps a number of operations per hour or day, usually related to a service plan or price. When these quotas are exceeded services must also provide immediate, actionable errors. Quotas and Limits should be scoped to a customer unit: a subscription, a tenant, an application, a plan, or without any other identification a range of ip addresses…as appropriate to the service goals so that the load is properly shared and one unit is not interfering with another.
Fixed

**Pull request merged**